clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -2163,6 +2163,49 @@ static void BuildFlattenedTypeList(QualType BaseTy,
   }
 }
 
+bool SemaHLSL::IsLineVectorLayoutCompatibleType(clang::QualType QT) {
+  if (QT.isNull())
+    return false;
+
+  llvm::SmallVector<QualType, 16> QTTypes;
+  BuildFlattenedTypeList(QT, QTTypes);
+
+  QualType FirstQT = QTTypes[0];
```

Comment:
Maybe we could add an assert on the vector size or a check with `llvm_unreachable` when size is 0